File: pkg_rospy/scripts/simple_action_client.py
#! /usr/bin/env python

# Sends goal and waits for result, similar to ordinary client-server setup

import rospy
import actionlib
from pkg_rospy.msg import CountNumbersAction, CountNumbersGoal

class CountNumbersClient():
    def __init__(self):
        self._ac = actionlib.SimpleActionClient('/count_numbers', CountNumbersAction)
        self._ac.wait_for_server()
        rospy.loginfo("connected to actionserver")

    def send_goal_get_result(self):
        goal = CountNumbersGoal()
        goal.count_upto = 10
        goal.wait_time = 1.0
        self._ac.send_goal(goal)
        rospy.loginfo("Goal has been sent")
        success = self._ac.wait_for_result(rospy.Duration(12.0))
        if not success:
            rospy.logwarn("Timeout")
        rospy.loginfo(self._ac.get_result())

def main():
    rospy.init_node("count_numbers_client")
    Client = CountNumbersClient()
    Client.send_goal_get_result()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: pkg_roscpp/src/simple_action_client.cpp
#include "ros/ros.h"
#include "actionlib/client/simple_action_client.h"
#include "pkg_roscpp/CountNumbersAction.h"

class ActionClient{
protected:
    ros::NodeHandle nh_;
    actionlib::SimpleActionClient<pkg_roscpp::CountNumbersAction> ac_;

public:
    ActionClient() : 
        ac_("count_numbers", true)
    {
        ROS_INFO("waiting for server to start");
        ac_.waitForServer();    
    }

    ~ActionClient(void){}

    void send_goal(){
        ROS_INFO("connected to server");
        pkg_roscpp::CountNumbersGoal goal;
        goal.count_upto = 10;
        goal.wait_time = 1.0;
        ac_.sendGoal(goal, 
                boost::bind(&ActionClient::done_CB, this, _1, _2)); //_1 _2 -> number of arguments in done_CB()
        ROS_INFO("goal sent");
    }

    void done_CB(const actionlib::SimpleClientGoalState &state, 
            const pkg_roscpp::CountNumbersResultConstPtr &result)
    {
        //ROS_INFO("Finished in state: %s", (state).toString.c_str());
        ROS_INFO("Result: [%d]", (int)result->final_count);
    }
};

int main(int argc, char **argv){
    ros::init(argc, argv, "simple_action_client");
    ActionClient Client;
    Client.send_goal();
    ros::spin();
    return 0;
}